### PR TITLE
Service patch 

### PR DESCRIPTION
**Pull request by Hammad887**
.env file and docker compose pushed

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    "configurations": [
+        {
+            "name": "Launch Package",
+            "type": "go",
+            "request": "launch",
+            "mode": "auto",
+            "program": "${fileDirname}"
+        }
+    ]
+}
```
Comment (shahzaibaziz) [inline .vscode/launch.json]: remove this

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	@echo
+	@echo 'Commands:'
+	@echo '  build           		Compile project.'
+	@echo '  check           		Run linters.'
+	@echo '  format          		Format source code.'
+	@echo '  codegen         		Generate code.'
+	@echo '  test            		Run test case'
+	@echo '  prepare         		build dev container'
+	@echo '  clean       			remove dev temp folder'
+	@echo
```
Comment (shahzaibaziz) [inline Makefile]: new line at the end

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+prepare: tmp/dev_image_id
+tmp/dev_image_id:
+	@mkdir -p tmp
+	@docker rmi -f ${DEV_IMAGE} > /dev/null 2>&1 || true
+	@docker build -t ${DEV_IMAGE} -f Dockerfile.dev .
+	@docker inspect -f "{{ .ID }}" ${DEV_IMAGE} > tmp/dev_image_id
+
+clean:
+	@${DOCKRUN} bash -c 'rm -rf bin tmp vendor'
+vendor: tmp/vendor-installed
```
Comment (shahzaibaziz) [inline Makefile]: remove this

File: .DS_Store
Comment (shahzaibaziz) [inline .DS_Store:1]: remove this

File: db/mysql/client.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	ctx := context.Background()
+
+	username := viper.GetString(config.MySQLDBUsername)
+	password := viper.GetString(config.MySQLDBPassword)
+	hostname := viper.GetString(config.MySQLDBHost)
+	dbName := viper.GetString(config.MySQLDBName)
+	// username := "hammad"
+	// password := "Hammad_887"
+	// hostname := "127.0.0.1:3306"
+	// dbName := "chatapp"
```
Comment (shahzaibaziz) [inline db/mysql/client.go]: remove

File: db/mysql/client.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+	fmt.Println(dsn(username, password, hostname, dbName))
+
+	// dbClient, err := sql.Open("mysql", "abc:password@tcp(db:3306)/test_database")
+
+	dbClient, err := sql.Open("mysql", "hammad:Hammad_887@tcp(localhost:3306)/chatapp")
```
Comment (shahzaibaziz) [inline db/mysql/client.go]: use environment variable here to connect with database

File: db/mysql/ops.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+	"database/sql"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	domain "github.com/Hammad887/chat-app/models"
```
Comment (shahzaibaziz) [inline db/mysql/ops.go]: arrange imports

File: db/mysql/ops.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+		log.Println(err)
+		return false, err
+	}
+	return true, nil
+}
+
+func (c *client) assignUserToChatRoom(user *domain.User, chatroom *domain.ChatRoom) error {
+	// vars := mux.Vars(r)
+	roomID := chatroom.ID
+	userID := user.ID
```
Comment (shahzaibaziz) [inline db/mysql/ops.go]: don't create extra variable here

File: db/client.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+func Register(name string, factory DataStoreFactory) {
+	if factory == nil {
+		log.Fatalf("Datastore factory %s does not exist.", name)
+		return
+	}
+	if _, ok := datastoreFactories[name]; ok {
+		log.Fatalf("Datastore factory %s already registered. Ignoring.", name)
+		return
+	}
+	datastoreFactories[name] = factory
```
Comment (shahzaibaziz) [inline db/client.go]: please explain this

File: db/mysql/client.go (hunk 5 above)
Comment (shahzaibaziz) [inline db/mysql/client.go]: use environment variable here to connect with database

File: config/config.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+func init() {
+	viper.SetConfigType("env")
+	viper.SetConfigFile(".env") // Assuming you're using .env in the root
+	viper.AutomaticEnv()
+
+	err := viper.ReadInConfig()
+	fmt.Println(viper.AllSettings())
+
+	if err != nil {
+		fmt.Printf("Error reading config file: %s\n", err)
```
Comment (shahzaibaziz) [inline config/config.go]: use log package

File: Dockerfile.dev [hunk 10]
```diff
@@ -0,0 +1,23 @@
+FROM golang:1.18
```
Comment (shahzaibaziz) [inline Dockerfile.dev]: use the latest version

File: db/mysql/ops.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+	if _, err := c.dbc.Exec("INSERT INTO users (id, name, email, password) VALUES (?, ?, ?, ?)", user.ID, user.Name, user.Email, string(hashedPassword)); err != nil {
+		return false, err
+
+	}
+
+	// createChatRoom()
+	return true, nil
+}
+
+func (c *client) SendMessage(ctx context.Context, id string, message *domain.Message) (bool, error) {
```
Comment (shahzaibaziz) [inline db/mysql/ops.go]: why using bool here, you can achieve this without error return type only

Comment (shahzaibaziz) [inline db/mysql/ops.go]: why using bool here, you can achieve this without error return type only

File: db/mysql/ops.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+func extractTokenFromRequest(r *http.Request) string {
+	authHeader := r.Header.Get("Authorization")
+	if authHeader != "" {
+		splitToken := strings.Split(authHeader, "Bearer ")
+		if len(splitToken) == 2 {
+			return splitToken[1]
+		}
+	}
+	return ""
+}
```
Comment (shahzaibaziz) [inline db/mysql/ops.go]: this logic should moe to service layer

File: db/mysql/ops.go [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+
+	tokenString, err := token.SignedString(jwtKey)
+	if err != nil {
+		return "", err
+	}
+
+	return tokenString, nil
+}
+
+func tokenVerificationMiddleware(next http.Handler) http.Handler {
```
Comment (shahzaibaziz) [inline db/mysql/ops.go]: this should be part of api handler layers

File: docker-compose.yaml [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+      MYSQL_ROOT_PASSWORD: "password"
+      MYSQL_DATABASE: "chatapp"
+      MYSQL_USER: "newuser"
+      MYSQL_PASSWORD: "password"
+      MYSQL_ROOT_HOST: '%' 
+      # MYSQL_DB_HOSTS: mysql:3306
+      # MYSQL_DB_NAME: userdb
+      # MYSQL_DB_USERNAME: root
+      # MYSQL_DB_PASSWORD: password
+    # command: --default-authentication-plugin=mysql_native_password
```
Comment (shahzaibaziz) [inline docker-compose.yaml]: remove

File: docker-compose.yaml [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+  swagger-ui:
+    image: swaggerapi/swagger-ui
+    container_name: "swagger-ui"
+    ports:
+      - "8001:8080"
+    volumes:
+      - ./swagger.yaml:/swagger.yaml
+    environment:
+      SWAGGER_JSON: /swagger.yaml
```
Comment (shahzaibaziz) [inline docker-compose.yaml:46]: explain why using this here

File: docs/client/chatroom_client.go
Comment (shahzaibaziz) [inline docs/client/chatroom_client.go:1]: name this docs to gen

Comment (shahzaibaziz) [inline docs/client/chatroom_client.go:1]: explain this purpose of this module

File: handlers/handler.go [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+func NewCustomHandler(api *operations.ChatroomAPI, rt *runtime.Runtime) {
+	api.ServiceRegisterUserHandler = RegisterUserHandler(rt)
+	api.ServiceLoginHandler = LoginUserHandler(rt)
+	api.ServiceLogoutHandler = LogoutUserHandler(rt)
+	api.ServiceGetAllChatroomsHandler = GetAllChatroomsHandler(rt)
+	api.ServiceGetChatroomHandler = GetChatroomHandler(rt)
+	api.ServiceSendMessageHandler = SendMessageHandler(rt)
+	api.ServiceGetAllMessagesHandler = GetChatroomMessagesHandler(rt)
+	// add all new handlers
```
Comment (shahzaibaziz) [inline handlers/handler.go:16]: you can use bearer token protocol to authenticate request

File: handlers/logger.go [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+)
+
+func log(ctx context.Context) logrus.FieldLogger {
+	return logrus.WithContext(ctx).WithField("pkg", "handler")
+}
```
Comment (shahzaibaziz) [inline handlers/logger.go:11]: you can use this same functionality in service and database layers

File: handlers/registerHandler.go [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+	rt *runtime.Runtime
+}
+
+// Handle implements service.RegisterUserHandler.
+func (r *registerUser) Handle(params service.RegisterUserParams) middleware.Responder {
+	ctx := context.Background()
+
+	user := params.User
+
+	success, err := r.rt.Service().RegisterUser(context.Background(), &genModel.User{
```
Comment (shahzaibaziz) [inline handlers/registerHandler.go]: covert gen model to domain model in handler and always pass domain model in the service layer

File: migration.sql
Comment (shahzaibaziz) [inline migration.sql:1]: explain the purpose of this file

File: runtime.go [hunk 19]
```diff
@@ -0,0 +1,29 @@
+package runtime
+
+import (
+	"github.com/Hammad887/chat-app/db/mysql"
+	wraperrors "github.com/pkg/errors"
+
+	"github.com/Hammad887/chat-app/db"
+	"github.com/Hammad887/chat-app/service"
```
Comment (shahzaibaziz) [inline runtime.go]: sort

File: scripts/swagger.sh [hunk 20]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+if [ ! -d "docs" ]; then
+  mkdir docs 
+fi
+
+swagger generate server -t docs --exclude-main
```
Comment (shahzaibaziz) [inline scripts/swagger.sh]: use gen instead of docs

File: service/helper.go
Comment (shahzaibaziz) [inline service/helper.go:1]: this should be a part of handler layers

File: service/interface.go [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"context"
+
+	"github.com/Hammad887/chat-app/db"
+	genModel "github.com/Hammad887/chat-app/docs/models"
+)
+
+// Manager defines the available functions for the given service implementation.
+type Manager interface {
+	GetChatroom(ctx context.Context, id string) (*genModel.Chatroom, error)
```
Comment (shahzaibaziz) [inline service/interface.go]: nly use domain model

File: service/users.go [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+	messages, err := s.db.GetChatroomMessages(ctx, id)
+	if err != nil {
+		return nil, wraperrors.Wrap(err, "failed to find user with given id")
+	}
+
+	return asMessagesResponse(messages), nil
+}
+
+// GetChatrooms implements Manager.
+func (s *service) GetChatrooms(ctx context.Context) ([]*genModel.Chatroom, error) {
```
Comment (shahzaibaziz) [inline service/users.go]: this function name should be ListChatRoom

**Review by shahzaibaziz — CHANGES_REQUESTED** [review]

File: .vscode/launch.json (hunk 1 above)
Comment (Hammad887) [reply .vscode/launch.json]: Removed

File: Makefile (hunk 2 above)
Comment (Hammad887) [reply Makefile]: Added the new line at the end

File: .DS_Store
Comment (Hammad887) [reply .DS_Store:1]: Removed

File: db/mysql/client.go (hunk 4 above)
Comment (Hammad887) [reply db/mysql/client.go]: Removed

File: db/mysql/ops.go (hunk 6 above)
Comment (Hammad887) [reply db/mysql/ops.go]: Fixed. Used goimports to arrange the imports

File: docker-compose.yaml (hunk 14 above)
Comment (Hammad887) [reply docker-compose.yaml]: Removed

File: runtime.go (hunk 19 above)
Comment (Hammad887) [reply runtime.go]: Sorted

File: Makefile (hunk 3 above)
Comment (Hammad887) [reply Makefile]: Removed

File: db/mysql/client.go (hunk 5 above)
Comment (Hammad887) [reply db/mysql/client.go]: Done

File: db/client.go (hunk 8 above)
Comment (Hammad887) [reply db/client.go]: Was referencing external source. Has been removed now.

File: db/mysql/ops.go (hunk 7 above)
Comment (Hammad887) [reply db/mysql/ops.go]: Removed

File: docker-compose.yaml (hunk 15 above)
Comment (Hammad887) [reply docker-compose.yaml:46]: When running the app in container, I want the rest api to be useable.

File: migration.sql
Comment (Hammad887) [reply migration.sql:1]: To create the db environment for anyone that wants to run the app.

File: config/config.go (hunk 9 above)
Comment (Hammad887) [reply config/config.go]: Done.

File: db/mysql/client.go (hunk 5 above)
Comment (Hammad887) [reply db/mysql/client.go]: Changed to environment variables.

File: Dockerfile.dev (hunk 10 above)
Comment (Hammad887) [reply Dockerfile.dev]: Done

File: db/mysql/ops.go (hunk 11 above)
Comment (Hammad887) [reply db/mysql/ops.go]: My bad. You are absolutely right. I have removed the bool.

Comment (Hammad887) [reply db/mysql/ops.go]: Changed

File: db/mysql/ops.go (hunk 12 above)
Comment (Hammad887) [reply db/mysql/ops.go]: Done

File: db/mysql/ops.go (hunk 13 above)
Comment (Hammad887) [reply db/mysql/ops.go]: Removed from the ops.go in db layer

File: docs/client/chatroom_client.go
Comment (Hammad887) [reply docs/client/chatroom_client.go:1]: Changed

Comment (Hammad887) [reply docs/client/chatroom_client.go:1]: Swagger auto-generated this client library for interacting with a chatroom service without directly making HTTP calls. I think it is used for ease of developers.

File: handlers/handler.go (hunk 16 above)
Comment (Hammad887) [reply handlers/handler.go:16]: My task was a bit different. It's requirement doesn't specify the authentication of token.

File: handlers/logger.go (hunk 17 above)
Comment (Hammad887) [reply handlers/logger.go:11]: Added

File: handlers/registerHandler.go (hunk 18 above)
Comment (Hammad887) [reply handlers/registerHandler.go]: Changed

File: scripts/swagger.sh (hunk 20 above)
Comment (Hammad887) [reply scripts/swagger.sh]: Edited the script. Changed to gen

File: service/helper.go
Comment (Hammad887) [reply service/helper.go:1]: Added to handler layer

File: service/interface.go (hunk 21 above)
Comment (Hammad887) [reply service/interface.go]: Changed

File: service/users.go (hunk 22 above)
Comment (Hammad887) [reply service/users.go]: Changed the function name

**Comment by Hammad887** [comment]
Done all the changes. Thank you for the reviews!

File: docker-compose.yaml (hunk 15 above)
Comment (shahzaibaziz) [reply docker-compose.yaml:46]: No, it does not connect to your server. We are using this to create a separate endpoint for API documentation. You will need to conduct some research on this.

File: handlers/handler.go (hunk 16 above)
Comment (shahzaibaziz) [reply handlers/handler.go:16]: are you saying your application does not support Token verification?

File: service/logger.go [hunk 23]
```diff
@@ -0,0 +1,11 @@
+package service
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+)
+
+func log(ctx context.Context) logrus.FieldLogger {
```
Comment (shahzaibaziz) [inline service/logger.go:9]: you are not using this

**Pull request merged**